core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRandomDistributionEndpointSelector.java

```diff
@@ -0,0 +1,152 @@
+            currentTotalWeightUpdater =
+            AtomicLongFieldUpdater.newUpdater(
+                    WeightedRandomDistributionEndpointSelector.class, "currentTotalWeight");
+
+    private final List<Entry> entries;
+    private final long totalWeight;
+    private volatile long currentTotalWeight;
+
+    WeightedRandomDistributionEndpointSelector(List<Endpoint> endpoints) {
+        final ImmutableList.Builder<Entry> builder = ImmutableList.builder();
+
+        long totalWeight = 0;
+        for (Endpoint endpoint : endpoints) {
+            if (endpoint.weight() <= 0) {
+                continue;
+            }
+            builder.add(new Entry(endpoint));
+            totalWeight += endpoint.weight();
+        }
+        this.totalWeight = totalWeight;
+        currentTotalWeight = totalWeight;
+        entries = builder.build();
+    }
+
+    @VisibleForTesting
+    List<Entry> entries() {
+        return entries;
+    }
+
+    @Nullable
+    Endpoint selectEndpoint() {
+        if (entries.isEmpty()) {
+            return null;
+        }
+        final ThreadLocalRandom threadLocalRandom = ThreadLocalRandom.current();
+        Endpoint selected = null;
+        for (;;) {
+            final long currentWeight = currentTotalWeight;
+            if (currentWeight == 0) {
+                // currentTotalWeight will become totalWeight as soon as it becomes 0 so we just loop again.
+                continue;
+            }
+            long nextLong = threadLocalRandom.nextLong(currentWeight);
+            // There's a chance that currentTotalWeight is changed before looping currentEntries.
+            // However, we have counters and choosing an endpoint doesn't have to be exact so no big deal.
+            // TODO(minwoox): Use binary search when the number of endpoints is greater than N.
+            for (Entry entry : entries) {
+                if (entry.isFull()) {
+                    continue;
+                }
+                final Endpoint endpoint = entry.endpoint();
+                final int weight = endpoint.weight();
+                nextLong -= weight;
+                if (nextLong < 0) {
+                    final int counter = entry.incrementAndGet();
+                    if (counter <= weight) {
+                        selected = endpoint;
+                        if (counter == weight) {
+                            decreaseCurrentTotalWeight(weight);
+                        }
+                    }
+                    break;
+                }
+            }
+
+            if (selected != null) {
+                return selected;
+            }
+        }
+    }
+
+    private void decreaseCurrentTotalWeight(int weight) {
+        for (;;) {
+            final long oldWeight = currentTotalWeight;
+            final long newWeight = oldWeight - weight;
+            if (currentTotalWeightUpdater.compareAndSet(this, oldWeight, newWeight)) {
```

Comment:
Does `addAndGet(this, -weight) == 0` without a spin work? This spin looks very suspicious at first glance (no chance of infinite spin?), if we can remove it then problem gone.

Reply:
That's a good suggestion.